### ISSUE TITLE
Engine's save and load state_dict don't use the full state

### ISSUE THREAD

**Issue by achigeor** [pytorch/ignite#741]
I have a question regarding the Engine's state and more specifically the engine's metrics.

When using `ModelCheckpoint()` to save an `Engine()` object, it only saves (according to docs):
* Iteration / epoch
* max_epochs
* epoch_length
* seed

This omits the metrics (if any). Accordingly, when loading an Engine's state, metrics are hardcoded as an empty dictionary https://github.com/pytorch/ignite/blob/master/ignite/engine/engine.py#L523.

Is there a specific reason that this happens? 

The usecase for saving metrics is that when we resume training, we need the metrics to be same as before, so we can correctly keep track of the best models we want to save. Is this currently possible only by manually saving and loading the metrics in the engine state or am I missing something? 

Thanks!

**Comment by vfdev-5** [comment]
@achigeor idea is to save only necessary thing to resume the training part only handled by the Engine. Metrics can be recomputed and anyway we do not store the history of them in `engine.state`. 

For example, when we train a network on MNIST (https://github.com/pytorch/ignite/tree/master/examples/mnist#training-save--resume) and the training is stopped (e.g. due to an exception) in the middle. Logging system like Tensorboard and checkpointer like `ModelCheckpoint` store computed metrics, engine state, model state_dict, optimizer state_dict params etc before the stop. 
When a new run is started by resuming from a stored state. Metrics can be again recomputed for the following epochs/iterations. As the curves (training/accuracy and validation accuracy) on the figure here: https://github.com/pytorch/ignite/blob/master/examples/mnist/assets/save_resume_p1.png 
But I agree when resuming the training we do not recover on the logging system previously computed metrics and IMO this does not make a lot of sense.

If you still would like to store somewhere the metrics and reload them. For example, we store them into a file. So this can be done manually like that:
```python
from pathlib import Path
import torch

trainer = ...
evaluator = ...

metrics_history_fp = Path("metrics_history.pth")

# Here we initialize (and reload) metrics history
@trainer.on(Events.STARTED)
def init(_):
    trainer.state.metrics_history = []
    if metrics_history_fp.exists():
        trainer.state.metrics_history = torch.load(metrics_history_fp.as_posix())

# Here we compute metrics and store them into the file
@trainer.on(Events.EPOCH_COMPLETED)
def compute_metrics(_):
    state = evaluator.run(val_dataloader) 
    m = dict(state.metrics)
    m['epoch'] = trainer.state.epoch
    trainer.state.metrics_history.append(m)
    torch.save(trainer.state.metrics_history, 
                      metrics_history_fp.as_posix())

``` 


HTH

**Comment by vfdev-5** [comment]
@achigeor I close the issue as answered, please feel free to reopen if more support needed on this topic.
